### ISSUE TITLE
Can you give the training data file and the training output file (used by Post_processing.ipynb)?

### ISSUE THREAD

**Issue by MrChencc** [snwagh/falcon-public#4]
As used in the picture：

1、Post_processing.ipynb：
![image](https://user-images.githubusercontent.com/34539085/95653162-00ce0b00-0b29-11eb-9a25-6235585a9174.png)

2、secondary.cpp
![image](https://user-images.githubusercontent.com/34539085/95653177-2c50f580-0b29-11eb-9d57-2c9670774a15.png)

**Comment by snwagh** [comment]
@MrChencc, the `files/train_data_*` and others contain just a few MNIST samples are were used only for debugging purposes and won't be of much use. To reproduce the correctness runs, I've provided pre-trained data for SecureML network architecture (in `files/preload/SecureML/`). Once you run the MPC inference, the `SecureML.txt` file will be output by the code. You need to path it correctly to the `Post_processing.ipynb` script. 

In general, you can take a pre-trained network and use/modify the code in `secondary.cpp` appropriately to import the pre-trained model as well as a sample batch size. You can then use the same data in the plaintext model trained on any platform such as TF or PyTorch and compare its output with the output of the MPC. The `Post_processing.ipynb` is a small script to count the number of misclassifications after appropriate fixed-point to float conversions. If you need to reproduce the results for the other networks as well, I can try to retrieve the data from the servers.

**Comment by MrChencc** [comment]
@snwagh ,

Thank you very much, it helped me a lot.

And I hope you can give other pre-trained network data like `files/preload/SecureML/`.

Thanks again.

**Comment by snwagh** [comment]
Good question, this is to be implemented I see the comment in the code also indicates likewise. I won't have the time to implement this but I can give you pointers and you can submit a merge request once you have a working prototype. 

You will need to specify the paths from where each party receives its inputs (shares of training data, training label, testing data, testing label). You can tweak the paths but it would be better if you can use the reference code in function `preload_network` ([here](https://github.com/snwagh/falcon-public/blob/master/src/secondary.cpp#L138-L148)) to receive each party's share. The [SecureNN codebase](https://github.com/snwagh/securenn-public/tree/master/mnist) might have something useful for this, with tweaking for RSS, might be useful in parsing real data into simple shares (for instance (data, 0, 0) format in RSS).

The pre-trained models provided in the repo are generated are done using PyTorch (for instance [here](https://github.com/snwagh/falcon-public/blob/master/scripts/MNIST_SecureML.ipynb)). To actually train the model in C++, you will have to run the train function (after you fix the data importing). Once again, it won't have very high accuracy out of the box but anything about 30% should show non-trivial leaning and be used as an indication that things are working. With a bit more tweaking you should be able to get higher accuracies. 

To test the preloaded networks, just uncomment these lines ([here](https://github.com/snwagh/falcon-public/blob/master/src/main.cpp#L55-L56)), remove the training lines and add the testing lines in ([here](https://github.com/snwagh/falcon-public/blob/master/src/main.cpp#L73-L78)).

Finally, I will try to add the other pretrained data as well. If I don't do it within a week, please go ahead and reopen this issue.

**Comment by MrChencc** [comment]
Okay, thanks for the reply, I will try to experiment next.

**Comment by snwagh** [comment]
@MrChencc, I've added the other pre-trained data [(here)](https://github.com/snwagh/falcon-public/tree/master/files/preload), this should help you reproduce the other results as well.
